search/types.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package search
+
+import (
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/postings"
+
+	"github.com/m3db/m3x/errors"
+)
+
+// Query is a search query for documents.
+type Query interface {
+	// Searcher returns a Searcher for executing the query over a collection of Readers.
+	Searcher(rs index.Readers) (Searcher, error)
```

Comment:
talked offline: change to accept snapshot

Reply:
Thought about this a little more, I'm not sure we want to pass the `Snapshot` here because then the lifecycle of the `Snapshot` becomes tied to the `Searcher`s. I think it would be preferable to have the layer above manage the `Snapshot` exclusively. That way it would know when the `Snapshot` is open or closed and be able to get `Reader`s from it as appropriate.

Reply:
Actually, changed my mind, I like passing the Snapshot because we only need a reference to it and not ownership, will update.

Reply:
Sounds good.